test/App/Fossa/Analyze/UploadSpec.hs

```diff
@@ -39,100 +43,212 @@ expectGetSuccess = do
+
+expectUploadFirstPartyDataToS3 :: Has MockApi sig m => m ()
+expectUploadFirstPartyDataToS3 = do
+  let mergedUnits = mergeSourceAndLicenseUnits Fixtures.sourceUnits Fixtures.firstLicenseSourceUnit
+  UploadFirstPartyScanResult Fixtures.signedUrl mergedUnits `returnsOnceForAnyRequest` ()
+
+expectedMergedFullSourceUnits :: NE.NonEmpty FullSourceUnit
+expectedMergedFullSourceUnits = NE.fromList [fullSourceUnit, fullLicenseUnit]
+  where
+    fullSourceUnit =
+      FullSourceUnit
+        { fullSourceUnitName = "testSourceUnitName"
+        , fullSourceUnitType = "testSourceUnitType"
+        , fullSourceUnitManifest = Just "testSourceUnitManifest"
+        , fullSourceUnitBuild = Nothing
+        , fullSourceUnitGraphBreadth = Complete
+        , fullSourceUnitOriginPaths = []
+        , fullSourceUnitAdditionalData = Nothing
+        , fullSourceUnitFiles = Nothing
+        , fullSourceUnitData = Nothing
+        , fullSourceUnitInfo = Nothing
+        }
+    fullLicenseUnit =
+      FullSourceUnit
+        { fullSourceUnitName = "empty"
+        , fullSourceUnitType = "LicenseUnit"
+        , fullSourceUnitManifest = Nothing
+        , fullSourceUnitBuild = Nothing
+        , fullSourceUnitGraphBreadth = Complete
+        , fullSourceUnitOriginPaths = []
+        , fullSourceUnitAdditionalData = Nothing
+        , fullSourceUnitFiles = Just $ "" NE.:| []
+        , fullSourceUnitData = Just $ emptyLicenseUnitData NE.:| []
+        , fullSourceUnitInfo = Just LicenseUnitInfo{licenseUnitInfoDescription = Nothing}
+        }
+
+uploadSuccessfulAnalysisSpec :: Spec
+uploadSuccessfulAnalysisSpec = do
+  describe
+    "uploadSuccessfulAnalysis"
+    $ do
+      baseDir <- runIO Fixtures.baseDir
+      it' "uploads analysis and git contributors"
+        . withGit mockGit
+        $ do
+          expectGetSuccess
+          expectAnalysisUploadSuccess
+          expectContributorUploadSuccess
+          locator <-
+            uploadSuccessfulAnalysis
+              baseDir
+              Fixtures.projectMetadata
+              (toFlag (JsonOutput) False)
+              Fixtures.projectRevision
+              Fixtures.sourceUnits
+              Nothing
+          locator `shouldBe'` expectedLocator
+      -- Currently our StdOut logging just writes directly to StdOut, so this is
+      -- just checking it doesn't fail.  In the future we should extract that so
+      -- we can test it better.
+      it' "renders JSON output when requested"
+        . withGit mockGit
+        $ do
+          expectAnalysisUploadSuccess
+          expectContributorUploadSuccess
+          expectGetSuccess
+          locator <-
+            uploadSuccessfulAnalysis
+              baseDir
+              Fixtures.projectMetadata
+              (toFlag (JsonOutput) True)
+              Fixtures.projectRevision
+              Fixtures.sourceUnits
+              Nothing
+          locator `shouldBe'` expectedLocator
+      it' "aborts when uploading to a monorepo"
```

Comment:
I _think_ you could ignore this test if Jess's monorepo PR merges before this PR.

Reply:
I'll leave it for now so that things compile (it's mostly a whitespace change, but I also had to add another parameter to `uploadSuccessfulAnalysis` to make things work

I'll delete the test once I merge Jess's PR - I'm assuming that I'll get a merge conflict if I don't